File: cs/Problems/Day2.cs
namespace aoc24.Problems;

// https://adventofcode.com/2024/day/2
public sealed class Day2 : IProblem<int>
{
    public int Solve(string input) => ComputeSafeReactorReportsOptimized(input);

    private static int ComputeSafeReactorReportsOptimized(ReadOnlySpan<char> input)
    {
        int safeReports = 0;
        Span<int> levels = stackalloc int[8];

        var lineEnumerator = input.Split(InputReader.NewLine).GetEnumerator();
        while (lineEnumerator.MoveNext())
        {
            var line = input[lineEnumerator.Current];
            int numOfLevels = InputParser.ParseNumbers(levels, line, ' ');
            bool increasing = levels[0] < levels[1];

            safeReports++;

            for (int i = 1; i < numOfLevels; i++)
            {
                var (prev, current) = (levels[i - 1], levels[i]);
                int diff = increasing ? current - prev : prev - current;

                if (diff is < 1 or > 3)
                {
                    safeReports--;
                    break;
                }
            }
        }

        return safeReports;
    }

    private static int ComputeSafeReactorReports(string input)
    {
        int safeReports = 0;

        foreach (var line in input.Split(InputReader.NewLine))
        {
            var levels = line.Split(" ").Select(int.Parse).ToArray();
            bool increasing = levels[0] < levels[1];

            safeReports++;

            for (int i = 1; i < levels.Length; i++)
            {
                var (prev, current) = (levels[i - 1], levels[i]);
                int diff = increasing ? current - prev : prev - current;

                if (diff is < 1 or > 3)
                {
                    safeReports--;
                    break;
                }
            }
        }

        return safeReports;
    }
}

File: go/problems/day2.go
package problems

import (
	"strconv"
	"strings"

	"github.com/Meryovi/hobby-aoc-2024/go/shared"
)

type Day2 struct{}

func (d Day2) Solve(input string) int {
	r := 0

	for _, line := range strings.Split(input, shared.NewLine) {
		values := strings.Split(line, " ")
		levels := make([]int, len(values))

		for i, v := range values {
			levels[i], _ = strconv.Atoi(v)
		}

		inc := levels[0] < levels[1]
		r++

		for i := 1; i < len(levels); i++ {
			prev := levels[i-1]
			curr := levels[i]
			diff := 0

			if inc {
				diff = curr - prev
			} else {
				diff = prev - curr
			}

			if diff < 1 || diff > 3 {
				r--
				break
			}
		}
	}

	return r
}
